pkg/config/provider.go

```diff
@@ -1,29 +1,211 @@
+	// skipped. For example, to include "aws_shield_protection_group" into
+	// the generated resources, one can add "aws_shield_protection_group$".
+	// To include whole aws waf group, one can add "aws_waf.*" to the list.
+	// Defaults to []string{".+"} which would include all resources.
+	IncludeList []string
+
+	// Resources is a map holding resource configurations where key is Terraform
+	// resource name.
+	Resources map[string]*Resource
+
+	// resourceConfigurators is a map holding resource configurators where key
+	// is Terraform resource name.
+	resourceConfigurators map[string]ResourceConfiguratorChain
+}
+
+// A ProviderOption configures a Provider.
+type ProviderOption func(*Provider)
+
+// WithGroupSuffix configures GroupSuffix for resources of this Provider.
+func WithGroupSuffix(s string) ProviderOption {
+	return func(p *Provider) {
+		p.GroupSuffix = s
+	}
+}
+
+// WithShortName configures ShortName for resources of this Provider.
+func WithShortName(s string) ProviderOption {
+	return func(p *Provider) {
+		p.ShortName = s
+	}
+}
+
+// WithIncludeList configures IncludeList for this Provider.
+func WithIncludeList(l []string) ProviderOption {
+	return func(p *Provider) {
+		p.IncludeList = l
+	}
+}
+
+// WithSkipList configures SkipList for this Provider.
+func WithSkipList(l []string) ProviderOption {
+	return func(p *Provider) {
+		p.SkipList = l
+	}
+}
+
+// WithBasePackages configures BasePackages for this Provider.
+func WithBasePackages(b BasePackages) ProviderOption {
+	return func(p *Provider) {
+		p.BasePackages = b
+	}
+}
+
+// WithDefaultResourceFn configures DefaultResourceFn for this Provider
+func WithDefaultResourceFn(f DefaultResourceFn) ProviderOption {
+	return func(p *Provider) {
+		p.DefaultResourceFn = f
+	}
+}
+
+// NewProvider builds and returns a new Provider.
+func NewProvider(resourceMap map[string]*schema.Resource, prefix string, modulePath string, opts ...ProviderOption) *Provider {
+	p := &Provider{
+		ModulePath:              modulePath,
+		TerraformResourcePrefix: fmt.Sprintf("%s_", prefix),
+		GroupSuffix:             fmt.Sprintf(".%s.tf.crossplane.io", prefix),
+		ShortName:               fmt.Sprintf("tf%s", prefix),
+		BasePackages:            DefaultBasePackages,
+		DefaultResourceFn:       DefaultResource,
+
+		IncludeList: []string{
+			// Include all Resources
+			".+",
+		},
+		Resources: map[string]*Resource{},
+
+		resourceConfigurators: map[string]ResourceConfiguratorChain{},
+	}
+
+	for _, o := range opts {
+		o(p)
+	}
+
+	for name, terraformResource := range resourceMap {
+		if len(terraformResource.Schema) == 0 {
+			// There are resources with no schema, that we will address later.
+			fmt.Printf("Skipping resource %s because it has no schema\n", name)
+			continue
+		}
+		if matches(name, p.SkipList) {
+			fmt.Printf("Skipping resource %s because it is in SkipList\n", name)
+			continue
+		}
+		if !matches(name, p.IncludeList) {
```

Comment:
Didn't want to add this because this might be too verbose while working on a small subset of all available resources. I believe we might consider some sort of debug option while generating.